### PR TITLE
fix(sql): fix table reader leak in over partition by symbol column queries

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Let's assume the following table:
```sql
create table 'trades' (
  ts TIMESTAMP,
  symbol SYMBOL,
  price DOUBLE
) timestamp (ts) PARTITION BY DAY;

insert into 'trades' values
  (1, 'a', 1),
  (2, 'b', 2),
  (3, 'c', 3);
```

This PR aims to fix some problematic queries that may be seen on this table.

### Problematic query 1

```sql
select avg(price) over (PARTITION BY symbol) from trades;
```

This one led to an `AssertionError` generated by `SqlCodeGenerator`.

### Problematic query 2

```sql
select *, row_number() over (partition by symbol order by symbol) from trades;
```

This one led to an NPE thrown by `CachedAnalyticRecordCursorFactory`'s `RecordChain` which in its turn led to a leaked `TableReader`.

### And even more problematic queries

```sql
select row_number() over (partition by symbol order by symbol), * from trades;
select row_number() over (partition by symbol order by symbol) from trades;
```

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 71 / 72 (98.61%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|63|63|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 72 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|63|63|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|4|4|100.00%|

**Review by ideoma — APPROVED** [review]

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -2322,7 +2322,7 @@ private RecordCursorFactory generateSelectAnalytic(QueryModel model, SqlExecutio
         // not main metadata to avoid partitionBy functions accidentally looking up
         // analytic columns recursively
 
-        // todo: these ar transient list, we can cache and reuse
+        // todo: these is a transient list, we can cache and reuse
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: Haha indeed. Fixed in d23060ba8ad8f3f7d5fe2c4af9f1fbc425743ce1

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 72 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|63|63|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|4|4|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java [hunk 2]
```diff
@@ -104,92 +106,156 @@ protected void _close() {
...
+            recordChain.toTop();
         }
 
-        recordChain.toTop();
-        return recordChain;
-    }
+        @Override
+        public void close() {
+            base.close();
+            recordChain.close();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java:218]: We're sinking down in this murky semantics that close() doesn't actually make object unusable.

RecordCursor is not "mutable", so it doesn't have clear "clear()" method. Instead it is sort of mutable with `toTop` call.

The semantics of what we do here is now a bit of spaghetti. The intent behind closing cursor and opening new one is that next open will re-use memory allocated by previous runs and thus avoid free/alloc calls.

`clear()` method will be appropriate here. But for that to make sense `chain` should not be owned by the cursor, but rather factory. This is how all the reusable state is held, or should be held

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java (hunk 2 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java:218]: Going to address this later once I fix variations of the problematic query as @ideoma pointed out. These are:
```sql
select row_number() over (partition by symbol order by symbol), * from trades;
select row_number() over (partition by symbol order by symbol) from trades;
```

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java]: Tried to address the suggestion in f6fdca3e9ea34a02ae499cc52612ef788bd62f3d

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 82 / 82 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|8|8|100.00%|

**Review by bluestreak01 — APPROVED** [review]
I also resolved dangling to-dos here: https://github.com/questdb/questdb/pull/2282/commits/25f929efff07a8ad6100ac41243f86fcca3123d6

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 89 / 89 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|61|61|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|8|8|100.00%|

**Comment by puzpuzpuz** [comment]
> I also resolved dangling to-dos here: [25f929e](https://github.com/questdb/questdb/commit/25f929efff07a8ad6100ac41243f86fcca3123d6)

Thanks. LGTM

**Comment by puzpuzpuz** [comment]
@ideoma thanks for pointing me at the additional problematic queries!

**Pull request merged**